websearch/websearch/src/durability.rs

```diff
@@ -0,0 +1,522 @@
+                        drop(session);
+                    });
+                }
+                Some(DurableSearchSessionState::Replay { .. }) => {
+                    // Nothing special to clean up for replay state
+                }
+                None => {}
+            }
+        }
+    }
+
+    impl<Impl: ExtendedwebsearchGuest> GuestSearchSession for DurableSearchSession<Impl> {
+        fn next_page(&self) -> Result<Vec<SearchResult>, SearchError> {
+            let durability = Durability::<Vec<SearchResult>, SearchError>::new(
+                "golem_websearch",
+                "next_page",
+                DurableFunctionType::ReadRemote,
+            );
+
+            if durability.is_live() {
+                let mut state = self.state.borrow_mut();
+                let (result, new_live_session) = match &mut *state {
+                    Some(DurableSearchSessionState::Live { session }) => {
+                        let result =
+                            with_persistence_level(PersistenceLevel::PersistNothing, || {
+                                session.next_page()
+                            });
+
+                        match result {
+                            Ok(value) => {
+                                let persisted_result =
+                                    durability.persist(NoInput, Ok(value.clone()))?;
+                                (Ok(persisted_result), None)
+                            }
+                            Err(error) => {
+                                let _ = durability.persist::<_, Vec<SearchResult>, SearchError>(
+                                    NoInput,
+                                    Err(error.clone()),
+                                );
+                                (Err(error), None)
+                            }
+                        }
+                    }
+                    Some(DurableSearchSessionState::Replay {
+                        original_params,
+                        current_page,
+                        finished,
+                    }) => {
+                        *current_page += 1;
+
+                        if *finished {
+                            let empty_result = durability.persist(NoInput, Ok(Vec::new()))?;
+                            (Ok(empty_result), None)
+                        } else {
+                            // Reconstruct session for the current page
+                            let session = Impl::unwrapped_search_session(original_params.clone())?;
+                            let mut last_results = Vec::new();
+                            for _ in 0..*current_page {
+                                last_results = session.next_page()?;
+                            }
+                            // Check if this is the last page (no more results)
+                            if last_results.is_empty() {
+                                *finished = true;
```

Comment:
This is not correct, you need to rely on exactly same logic to compute finished as it's done internally in the providers.